### DISCUSSION TITLE
Mixing <const> with <comptime>

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#81] [General]
Based on #80 comment I made about the generated C code, I've got curious to see how C++ would behave with `constexpr` in this case, as assigning `<const>` to `<const>` resembles this behavior.

In C++ we can do the following:

```c++
#include <iostream>

static const int x = 1;
static constexpr int a = x;

int main()
{
    std::cout << a << '\n'; // outputs '1'
}
```
and works as expected.

I have tried the same in Nelua and it does not work...

```lua
local x <const> = 1
local a <comptime> = x
print(a) -- I was hoping to get '1' as my output
```

So, how exactly `<const>` getting assigned to another `<const>` works behind the scenes during preprocessing procedure?

**Comment by edubart** [comment]
This is not possible, because `<const>` is a runtime constant, while `<comptime>` is a compile-time constant. The compiler does not store or process values about runtime constants, even if it's visible, so they cannot be mixed, for this case to work both should be `<comptime>` instead.

> So, how exactly <const> getting assigned to another <const> works behind the scenes during preprocessing procedure?

The `<const>` just serves one purpose, to disallow variable assignment throwing a compile error on any assignment attempt, other than that, it behaves just as usual non-const variable. Their values are unknown at compile-time (like runtime variables values are unknown).

**Reply by stefanos82** [reply]
> This is not possible, because `<const>` is a runtime constant, while `<comptime>` is a compile-time constant. The compiler does not store or process values about runtime constants, even if it's visible, so they cannot be mixed, for this case to work both should be `<comptime>` instead.

When you say "runtime constant" here, you mean Nelua's runtime or executable's runtime process?

**Reply by edubart** [reply]
I mean executable's runtime process.
